client-spark/common/src/main/java/org/apache/spark/shuffle/celeborn/SendBufferPool.java

```diff
@@ -36,23 +37,23 @@ public static SendBufferPool get(int capacity) {
   private final int capacity;
 
   // numPartitions -> buffers
-  private LinkedList<byte[][]> buffers;
+  private final LinkedList<byte[][]> buffers;
 
   public SendBufferPool(int capacity) {
     this.capacity = capacity;
     buffers = new LinkedList<>();
   }
 
   public synchronized byte[][] acquireBuffer(int numPartitions) {
-    for (int i = 0; i < buffers.size(); i++) {
-      if (buffers.get(i).length == numPartitions) {
-        return buffers.remove(i);
+    Iterator<byte[][]> iterator = buffers.iterator();
+    while (iterator.hasNext()) {
+      byte[][] candidate = iterator.next();
+      if (candidate.length == numPartitions) {
+        iterator.remove();
+        return candidate;
       }
     }
-    if (buffers.size() == capacity) {
-      buffers.removeFirst();
```

Comment:
what's the benefit of shrink buffer during acquire? it should be ok to do it only in return

Reply:
The purpose is to make sure total number of buffers will not exceed capacity. Say we don't have matching buffer in the pool, and buffers.size() equals capacity, then acquireBuffer will return null, HashBasedShuffleWriter will new a new buffer, thus the total buffers will be ```capacity + 1```. For N tasks, total buffer number can be ```capacity + N```.

Reply:
with this eager release, the worst case could be `capacity + N - 1`, right?

Reply:
> with this eager release, the worst case could be `capacity + N - 1`, right?

I think you are right, and it's also not what we want. I think we should always try to remove if ```acquireBuffer``` returns no match.

Reply:
>  I think we should always try to remove if acquireBuffer returns no match.

Make sense, changed